nixos/multipath: add module for multipath-tools package

limeytexan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The multipath-tools package has existed in Nixpkgs for some time but
without a nixos module to configure/drive it. This patch creates a module
and updates the package to work in stage-1 boot.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

limeytexan committed:
nixos/multipath: add multipath module

The multipath-tools package had existed in Nixpkgs for some time but
without a nixos module to configure/drive it. This module provides
attributes to drive the majority of multipath configuration options
and is being successfully used in stage-1 and stage-2 boot to mount
/nix from a multipath-serviced iSCSI volume.

Credit goes to @grahamc for early contributions to the module and
authoring the NixOS module test.

grahamc commented on pkgs/os-specific/linux/multipath-tools/stage1.patch:
```diff
@@ -0,0 +1,60 @@
+diff --git a/multipathd/main.c b/multipathd/main.c
```
Attempted to upstream in https://github.com/opensvc/multipath-tools/pull/17

grahamc committed:
boot.initrd.extraFiles: allow users to add arbitrary files to the initrd

grahamc committed:
iscsi root initiator: allow specifying arbitrary iscsi commands too

grahamc reviewed (commented):
I submitted a PR against your branch to add a test and somewhat refactor the configuration a bit:

https://github.com/flox/nixpkgs/pull/1

limeytexan commented on nixos/modules/services/networking/multipath.nix:
```diff
@@ -0,0 +1,575 @@
+    # with the `-1` argument which disables systemd calls. Invoke `multipath`
+    # to display the multipath mappings in the output of `journalctl -b`.
+    boot.initrd.kernelModules = [ "dm-multipath" "dm-service-time" ];
+    boot.initrd.secrets = {
+      "/etc/multipath/wwids" = "/etc/multipath/wwids";
+    };
```
Thanks for this. I think it would be fair to say this:

* _if_ multipath is required in stage-1 (i.e. for mounting `/` or `/nix`) then `/etc/multipath` will need to appear in the initrd - it won't work any other way
* otherwise if multipath is not required in stage-1 then there's no reason to embed `/etc/multipath` secrets in the initrd

With that, should we create a module `stage1Enable` boolean or similar, and then base the multipath invocations and `boot.initrd.secrets` additions based on the value of that variable? Also, to your point about the possibility of keeping secrets elsewhere, perhaps we solve that by creating a module `wwidsPath` variable or similar that defaults to `/etc/multipath/wwids`?

limeytexan committed:
multipath-tools: deprecate systemd-udev-settle.service

The systemd service manifest provided with the multipath-tools source
makes reference to the systemd-udev-settle.service, and this in turn
triggers deprecation warnings on startup. This patch removes these
references to silence these warnings.

cole-h reviewed (approved):
LGTM; I ran the test locally, and it passed :)

grahamc commented:
@ofborg eval
